src/python/pants/backend/python/lint/black/rules.py

```diff
@@ -44,24 +45,30 @@ async def _run_black(
         Digest, MergeDigests((request.snapshot.digest, config_files.snapshot.digest))
     )
 
-    result = await Get(
-        ProcessResult,
-        VenvPexProcess(
-            black_pex,
-            argv=(
-                *(("--config", black.config) if black.config else ()),
-                "-W",
-                "{pants_concurrency}",
-                *black.args,
-                *request.files,
+    with temporary_dir() as black_cache_dir:
```

Comment:
Hm, looking at how this `temporary_dir` function is used elsewhere, I don't see any other examples of using it rule code like this: mostly tests and some build-time helpers.

I think using it rule code is potentially slightly problematic too, as it does synchronous IO to create and clean-up the directory, which will block concurrent execution of other rules (~all IO should happen via `await`s and `Get`s etc.).

So, some other options:

1. find a Pants-y way to create a temporary dir in rule code
2. leverage the sandboxes being temporary directories, and using a subpath of that, e.g. `{"BLACK_CACHE_DIR": "__pants_black_cache_dir"}` (there seems to be a bit convention for using double underscores for non-user-created (either code or codegen) files that Pants has to synthesise into a sandbox)
3. something else?

Option 2 might also allow testing this more precisely, if we wanted to... although we could probably get away without this.

Reply:
Got it! Understood why we wouldn't want to use the temporary dir here, though that is a shame because the utility does a really nice job of cleaning up after itself.

What I can't work out is how to retrieve the sandbox dir from inside the Black rules, unless that `__pants_black_cache_dir` variable automatically creates a directory inside the sandbox.

Reply:
At this point of the rule code, the sandbox doesn't necessarily exist. Rules are conceptually configuring processes (arguments, environment variables, which files should be on disk) but not necessarily actually putting files on disk or executing anything: the results may be served from cache. 

The "files on disk" part is the `input_digest` argument, and the files and directories that implies are materialised within the sandbox if the process actually needs to run. This digest can be manipulated in various ways, including synthesising things out of thin air: the docs in https://www.pantsbuild.org/2.19/docs/writing-plugins/the-rules-api/file-system seem to be good background. For this particular case:

- new digest with an (empty) directory can be created via the `CreateDigest` rule (https://www.pantsbuild.org/2.19/docs/writing-plugins/the-rules-api/file-system#createdigest-create-new-files)
- multiple digests can be combined with `MergeDigests` (https://www.pantsbuild.org/2.19/docs/writing-plugins/the-rules-api/file-system#mergedigests-merge-collections-of-files)

Example:

https://github.com/pantsbuild/pants/blob/c15e54e5c8100a7d8638589bc6f675a95e7c49e9/src/python/pants/backend/scala/compile/scalac.py#L185-L186

However, for this particular case, I think black is resilient to the directory not existing and creates it itself (e.g. `rm -rf dir-that-doesnt-exist; BLACK_CACHE_DIR= dir-that-doesnt-exist black ...` seems to work for me, and creates `dir-that-doesnt-exist`), so either option likely works:

1. create the directory in advance with `CreateDigest` + `MergeDigests` so black doesn't have to
2. let black manage it internally, creating the directory if it needs to

2 seems simpler, so I'm inclined to that.

Reply:
For 2, we can't guarantee that the directory will be inside the sandbox without calling `CreateDigest` right? We'd probably have to choose a more global path like `/tmp/black-cache` or something like that?

Reply:
I suspect passing `BLACK_CACHE_DIR` as a relative path with no other set-up is likely to work fine: 

- when `BLACK_CACHE_DIR` is a relative path, it seems like black treats it as relative to its working directory (the test above creates `dir-that-doesnt-exist` in whatever directory I'm running that test in).
- pants executes this sort of process with the sandbox (or some subpath) as the working directory.

Are you observing something different?

Reply:
For whatever reason I can't get Pants/Black to create this folder. Here's what I ran:

```
./pants --no-local-cache --keep-sandboxes=always test src/python/pants/backend/python/lint/black::


18:28:03.44 [INFO] Initialization options changed: reinitializing scheduler...
18:28:36.09 [INFO] Scheduler initialized.
18:28:39.87 [INFO] Preserving local process execution dir /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-QTtkKX for Find interpreter for constraints: CPython==3.9.*
18:28:42.08 [INFO] Preserving local process execution dir /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-joyWaA for Building 12 requirements for requirements.pex from the 3rdparty/python/user_reqs.lock resolve: PyYAML<7.0,>=6.0, ansicolors==1.1.8, packaging==21.3, pex==2.2.1, pytest<7.1.0,>=6.2.4, python-lsp-jsonrpc==1.0.0, setuptools<64.0,>=63.1.0, toml==0.10.2, types-PyYAML==6.0.3, types-setuptools==62.6.1, types-toml==0.10.8, typing-extensions==4.3.0
18:28:42.15 [INFO] Preserving local process execution dir /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-jR7ayg for Building 8 requirements for pytest.pex from the 3rdparty/python/pytest.lock resolve: ipdb, pygments, pytest-asyncio, pytest-cov!=2.12.1,<3.1,>=2.12, pytest-html, pytest-icdiff, pytest-xdist<3,>=2.5, pytest==7.0.1
18:28:42.24 [INFO] Preserving local process execution dir /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-1LYeqb for Building 12 requirements for requirements.pex from the 3rdparty/python/user_reqs.lock resolve: PyYAML<7.0,>=6.0, ansicolors==1.1.8, packaging==21.3, pex==2.2.1, pytest<7.1.0,>=6.2.4, python-lsp-jsonrpc==1.0.0, setuptools<64.0,>=63.1.0, toml==0.10.2, types-PyYAML==6.0.3, types-setuptools==62.6.1, types-toml==0.10.8, typing-extensions==4.3.0
18:28:43.88 [INFO] Preserving local process execution dir /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-GIILTG for Building local_dists.pex
18:28:45.29 [INFO] Completed: Building local_dists.pex
18:28:45.29 [INFO] Preserving local process execution dir /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-eDrY12 for Searching for `bash` on PATH=/usr/bin:/bin:/usr/local/bin:/opt/homebrew/bin
18:28:45.45 [INFO] Preserving local process execution dir /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-o2rLoe for Test binary /bin/bash.
18:28:45.54 [INFO] Preserving local process execution dir /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-3TPezR for Building pytest_runner.pex
18:28:46.95 [INFO] Completed: Building pytest_runner.pex
18:28:46.96 [INFO] Preserving local process execution dir /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-vqcOys for Run Pytest for src/python/pants/backend/python/lint/black/rules_integration_test.py:tests
18:29:17.11 [INFO] Completed: Run Pytest - src/python/pants/backend/python/lint/black/rules_integration_test.py:tests - succeeded.

✓ src/python/pants/backend/python/lint/black/rules_integration_test.py:tests succeeded in 30.12s.
```

However, if I then try to find the cache dir from the top-level dir above each individual sandbox, I can't:

```
sudo find /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T -name __pants_black_cache_dir 2> /dev/null
```

(This command returns nothing)

Reply:
Ah, hm... I think the tests that pants runs are a bit weird, because there's two layers of sandbox going on:

1. the sandbox that the "outer" pants is using to run the test files
2. the sandboxes that are created while executing each test itself

And `--keep-sandboxes` on a test run will set that for 1, but not 2, I think. I don't know how to introspect 2 better.

What happens if you run `pants --keep-sandboxes=always fmt ...` to run black normally (i.e. not in a test)?

Reply:
Yup, that worked! Just for my own edification I tried both option 1 (https://github.com/pantsbuild/pants/pull/20638/commits/cda285f652894a87892a6530b48685352dad53f1) and option 2 (after reverting that commit) and both work properly.